bot/utils/persistent_scheduling.py

```diff
@@ -0,0 +1,155 @@
+        self._scheduler = Scheduler(name)
+        self.cache = RedisCache(namespace=name)
+
+        self.task_factory = task_factory
+
+        self._reschedule_task = event_loop.create_task(self.reschedule_all_tasks())
+
+    def __contains__(self, task_id: CacheKey) -> bool:
+        """Return True if a task with the given `task_id` is currently scheduled."""
+        return task_id in self._scheduler
+
+    async def wait_until_ready(self) -> None:
+        """Wait until the scheduler is ready and the cache is ready (and all cached tasks are scheduled)."""
+        await self._reschedule_task
+
+    async def get(self, task_id: CacheKey) -> t.Optional[Arrow]:
+        """Get the time in which `task_id`'s task should go off. If the task is not found return None."""
+        timestamp = await self.cache.get(task_id)
+
+        if not timestamp:
+            return None
+        return Arrow.utcfromtimestamp(timestamp)
+
+    async def to_dict(self) -> t.Dict[CacheKey, Arrow]:
+        """
+        Get a dict representation of the currently scheduled tasks.
+
+        The dictionary contains the task_id's as keys, and the times in which the tasks should go off as values.
+        """
+        dict_ = await self.cache.to_dict()
+        return {task_id: Arrow.utcfromtimestamp(timestamp) for task_id, timestamp in dict_.items()}
+
+    async def reschedule_task(self, task_id: CacheKey) -> None:
+        """Reschedule the task according to `task_factory` and `task_id`."""
+        time = await self.get(task_id)
+        if not time:
+            raise ValueError(f"Task with ID {task_id} was not found for rescheduling.")
+        self._scheduler.schedule_at(time.datetime, task_id, coroutine=self._to_schedule(task_id))
+
+    async def reschedule_all_tasks(self) -> None:
+        """Reschedule all tasks according to `task_factory` and the ID's in the cache."""
+        tasks = await self.cache.items()
+        for task_id, timestamp in tasks:
+            time = Arrow.utcfromtimestamp(timestamp)
+            self._scheduler.schedule_at(time.datetime, task_id, coroutine=self._to_schedule(task_id))
+
+    async def delete(self, task_id: CacheKey) -> None:
+        """
+        Unschedule the task identified by `task_id` and delete it from the cache.
+
+        `cancel` is left as-is in order to be able to pause execution until the next reschedule.
+        Log a warning if the task doesn't exist.
+        """
+        await self.cache.delete(task_id)
+
+        await self._reschedule_task  # Wait for rescheduling to finish to avoid race conditions.
+        self._scheduler.cancel(task_id)
+
+    async def delete_all(self) -> None:
+        """
+        Unschedule all known tasks and wipe the cache.
+
+        `cancel_all` remains untouched so that it can be used in instances such as cog unloads while preserving cache.
+        """
+        self._reschedule_task.cancel()
```

Comment:
Since it's in a coroutine, you can cancel the task and then await it to make sure it's cancelled before moving on to `cancel_all`.

Reply:
Alternatively, you could just call `self.cancel_call` which takes care of both of these steps. It seems fine to be able to move the cache clear before/after that call. Could also use `self.cancel` inside of `delete()`.